lib/puppet/type/archive.rb

```diff
@@ -122,9 +122,9 @@ def should_to_s(value)
   end
 
   newparam(:source) do
-    desc 'archive file source, supports http|https|ftp|file|s3 uri.'
+    desc 'archive file source, supports puppet|http|https|ftp|file|s3 uri.'
     validate do |value|
-      unless value =~ URI.regexp(%w(http https ftp file s3)) || Puppet::Util.absolute_path?(value)
+      unless value =~ URI.regexp(%w(puppet http https ftp file s3)) || Puppet::Util.absolute_path?(value)
         raise ArgumentError, "invalid source url: #{value}"
       end
```

Comment:
@hajee Perhaps you could declare a [provider feature](https://docs.puppet.com/guides/custom_types.html#provider-features)?  Something like...
```ruby
feature :puppet_urls, 'The ability to download from Puppet URLs' if value == 'puppet'
```
and then in the ruby provider
```ruby
has_feature :puppet_urls
```